File: 3rd-semester/object-oriented-programming/containers/csharp/src/Models/Container.cs
namespace ShipCargoSystem.Models;

/// <summary>
/// Abstract base class representing a shipping container.
/// Concrete subclasses define the cost calculation for their cargo type.
/// </summary>
public abstract class Container
{
    /// <summary>Gets the unique container identifier.</summary>
    public string Code { get; }

    /// <summary>Gets the destination port or city.</summary>
    public string Destination { get; }

    /// <summary>Initialises a Container.</summary>
    /// <param name="code">Unique container identifier.</param>
    /// <param name="destination">Destination port or city.</param>
    protected Container(string code, string destination)
    {
        Code        = code;
        Destination = destination;
    }

    /// <summary>Calculates the shipping cost for this container.</summary>
    /// <returns>Shipping cost in currency units.</returns>
    public abstract double GetCost();

    /// <inheritdoc/>
    public override string ToString() =>
        $"{GetType().Name}[code={Code}, destination={Destination}, cost={GetCost():F2}]";
}

File: 3rd-semester/object-oriented-programming/containers/java/src/main/java/com/university/oop/containers/model/Container.java
package com.university.oop.containers.model;

import java.util.logging.Logger;

/**
 * Abstract base class representing a shipping container.
 *
 * <p>Each container has a unique code, a destination port, and a shipping cost
 * calculated by the concrete subclass according to its cargo type.</p>
 */
public abstract class Container {

    private static final Logger LOGGER = Logger.getLogger(Container.class.getName());

    private final String code;
    private final String destination;

    /**
     * Constructs a Container.
     *
     * @param code        unique container identifier
     * @param destination destination port or city
     */
    protected Container(String code, String destination) {
        this.code = code;
        this.destination = destination;
    }

    /** @return unique container identifier */
    public String getCode() { return code; }

    /** @return destination port or city */
    public String getDestination() { return destination; }

    /**
     * Calculates the shipping cost for this container.
     *
     * @return shipping cost in currency units
     */
    public abstract double getCost();

    /** Logs the container details. */
    public void printInfo() {
        LOGGER.info(String.format("Code: %s | Destination: %s | Cost: %.2f",
                code, destination, getCost()));
    }

    @Override
    public String toString() {
        return String.format("Container[code=%s, destination=%s, cost=%.2f]",
                code, destination, getCost());
    }
}
